File: dsa_with_python/heap/priority_queue.py
from min_heap import min_heap

mh = min_heap()


def is_empty():
    return mh.is_empty()


def size():
    return mh.size()


def add(item):
    mh.insert(item)


def peek():
    if is_empty():
        return None
    return mh.get_min()


def element():
    if is_empty():
        raise Exception("Priority Queue is Empty!")
    return mh.get_min()


def poll():
    if is_empty():
        return None
    return mh.extract_min()


def remove():
    if is_empty():
        raise Exception("Priority Queue is Empty!")
    return mh.extract_min()


if __name__ == "__main__":
    add(10)
    add(20)
    add(15)
    print(peek())
    print(poll())
    print(peek())
    print(poll())
    print(peek())
    print(remove())
    print(poll())
    print(remove())

File: dsaWithJava/heap/PriorityQueue.java
package heap;

import java.util.NoSuchElementException;

public class PriorityQueue {
    MinHeap mh = new MinHeap();

    public boolean isEmpty() {
        return mh.isEmpty();
    }

    public int size() {
        return mh.size();
    }

    public void add(int item) {
        mh.insert(item);
    }

    public Integer peek() {
        if (isEmpty())
            return null;
        return mh.getMin();
    }

    public Integer element() {
        if (isEmpty())
            throw new NoSuchElementException();
        return mh.getMin();
    }

    public Integer poll() {
        if (isEmpty())
            return null;
        return mh.extractMin();
    }

    public int remove() {
        if (isEmpty())
            throw new NoSuchElementException();
        return mh.extractMin();
    }

    public static void main(String[] args) {
        PriorityQueue pq = new PriorityQueue();
        pq.add(10);
        pq.add(20);
        pq.add(15);
        System.out.println(pq.peek());
        System.out.println(pq.poll());
        System.out.println(pq.peek());
        System.out.println(pq.poll());
        System.out.println(pq.peek());
        System.out.println(pq.remove());
        System.out.println(pq.poll());
        System.out.println(pq.remove());
    }
}
